backend: Add websocket multiplexer

knrt10 opened the pull request:
This adds websocket multiplexer to the backend. Frontend now need to make a single websocket call to the backend. Once that connection is established with the backend it will send message to the backend with appropriate data. Backend will open multiple websockets and act as a proxy to the frontend. It will make request to k8s server and return data to frontend.

This also adds retry logic if the connection is broken between frontend and backend.

Fixes: #1802

## Testing

- Run backend `cd backend && go run ./cmd -dev -enable-dynamic-clusters`
- Run frontend `cd frontend && npm start` (we need this for userID of the session).
- Open postman application and use the API [here](https://www.postman.com/knrt1017/workspace/headlamp/ws-raw-request/6717608832943442e3cc99a7)
- It has message payload which you need to edit 
- 

```json
{
  "clusterID": "minikube", // update it with the cluster you have running
  "path": "/api/v1/pods",
  "query": "watch=true",
  "userId": "9872e4f36eaadb0e" // go to localstorage and copy `headlamp-userID` and paste here
}
```

- Click on the `Connect` button
<img width="1360" alt="Screenshot 2024-10-22 at 2 08 34 PM" src="https://github.com/user-attachments/assets/f1ede6b7-1788-470c-b984-d1c21739fc18">

- Once connected hit the send button.
<img width="1365" alt="Screenshot 2024-10-22 at 2 09 41 PM" src="https://github.com/user-attachments/assets/46a5a9ed-79a0-4f5a-821e-348bbc066ef0">

- You will see response from your k8s cluster

<img width="1365" alt="Screenshot 2024-10-24 at 12 42 28 PM" src="https://github.com/user-attachments/assets/6d27ba83-77f2-487c-8ae9-68ac1f19aa4c">

- Create an nginx pod `kubectl run nginx --image=nginx` and try other things like logs and exec

#### Logs

```json
{
  "clusterID": "minikube",
  "path": "/api/v1/namespaces/default/pods/nginx/log",
  "query": "follow=true&tailLines=100",
  "userId": "9872e4f36eaadb0e"
}
```

<img width="1351" alt="Screenshot 2024-10-24 at 12 45 18 PM" src="https://github.com/user-attachments/assets/36e3c38b-9c96-4676-961a-3690a76142c9">

#### Exec

```json
{
  "clusterID": "minikube",
  "path": "/api/v1/namespaces/default/pods/nginx/exec",
  "query": "command=ls&command=-l&stdout=true&stderr=true",
  "userId": "9872e4f36eaadb0e"
}
```

<img width="1346" alt="Screenshot 2024-10-24 at 12 43 45 PM" src="https://github.com/user-attachments/assets/1a4076e9-cf03-4b1e-ac65-4186a3b61667">

**The data is sent as binary and it is encoded to bas64 to parse as JSON response.**

#### Close connection

```json
{
  "clusterID": "minikube",
  "path": "/api/v1/pods",
  "userId": "9872e4f36eaadb0e",
  "data": "close"
}
```

<img width="1361" alt="Screenshot 2024-10-24 at 12 40 07 PM" src="https://github.com/user-attachments/assets/39e05eab-2e78-4602-aed9-0b4240aed6a5">

You can edit `path` as per your needs and check different output.

sniok commented:
Nice, tested it and it seems to work

There are a couple of crucial things missing though:

- The data coming in doesn't contain information about which request it is 
- There's no way to close the request (e.g. stop watching `/pods`)

knrt10 commented:
Thank you for the feedback @sniok. Awesome suggestions. I did forget to add closing the connection on request. Will add 

Regarding attaching, I did not try it. Will check why it’s not working.

illume commented:
I think we can delay any work on attaching to the container if there is a problem. Because we mainly want this to help with watching.

illume commented:
> Backend Code coverage changed from 61.6% to 61.6%. Change: 0% 😃.

Hrmm. Not sure why this is at 0, since there are tests... lol.

knrt10 commented:
@sniok @illume updated the `code + PR description + postman calls`. You can now do all k8s operation. Please try and let me know. Also added the state handling of webhook

sniok commented:
Thanks for the changes.

- I'm not sure why "type" is required? Isn't path enough to uniquely identify the operation? 

- The messages still don't have information about which request they belong to.

![image](https://github.com/user-attachments/assets/f8056e95-22a7-4b47-ad78-f2796a83615a)

Imagine we started watching two /pods from two different clusters (or two different ReplicaSets,etc) and we need to figure out where to put this update that we received from websocket. Currently we just get a Pod without any info where it came from

- Also I don't see how to close the connection?

knrt10 commented:
@sniok thank you for the detailed message. I have updated the description, screenshot, postman API and code. You will now have detailed information from where request is coming from. `type` as mentioned was not needed so has been removed. Not you can also close connection. Added usage in the description. Thanks

illume commented:
Thanks @knrt10 

(It looks like close can be used, and now a clusterID can be used to identify the response.)

Does it make sense to start implementing the frontend/ integration now? Will you do that?

Note I started an issue here about multi cluster e2e tests here:
- https://github.com/headlamp-k8s/headlamp/issues/2460

illume commented:
Thanks for the explanation @knrt10 

So the response returns (clusterID, path, query, and userID) as well as the data. This can be used to know from which websocket it comes from.

We discussed the possibility that multiple ones could be made with the same of those... in this case it will not work. I don't think we need to cover this now though. It could be solved in the future by adding a separate unique ID for each one. Again, I don't think this is necessary for our use cases... we can use (clusterID, path, query, and userID) as the unique key.

illume commented:
If @sniok you don't have any more comments or concerns, then we can merge this?

@knrt10 will begin frontend integration probably tomorrow, but we can merge this in first.  (please correct me if I'm wrong)

sniok commented:
looks good, thanks!
